File: core/src/main/java/dev/bee/kanjianki/core/study/HintVisibility.java
package dev.bee.kanjianki.core.study;

public final class HintVisibility {
    private final HintLevel level;
    private final boolean tracePathsVisible;
    private final boolean outlineVisible;
    private final boolean strokeNumbersVisible;
    private final boolean startDotsVisible;
    private final boolean strokeCountVisible;
    private final int visibleStrokeCount;

    public HintVisibility(
            HintLevel level,
            boolean tracePathsVisible,
            boolean outlineVisible,
            boolean strokeNumbersVisible,
            boolean startDotsVisible,
            boolean strokeCountVisible,
            int visibleStrokeCount
    ) {
        this.level = level == null ? HintLevel.TRACE : level;
        this.tracePathsVisible = tracePathsVisible;
        this.outlineVisible = outlineVisible;
        this.strokeNumbersVisible = strokeNumbersVisible;
        this.startDotsVisible = startDotsVisible;
        this.strokeCountVisible = strokeCountVisible;
        this.visibleStrokeCount = Math.max(0, visibleStrokeCount);
    }

    public HintLevel level() {
        return level;
    }

    public boolean tracePathsVisible() {
        return tracePathsVisible;
    }

    public boolean outlineVisible() {
        return outlineVisible;
    }

    public boolean strokeNumbersVisible() {
        return strokeNumbersVisible;
    }

    public boolean startDotsVisible() {
        return startDotsVisible;
    }

    public boolean strokeCountVisible() {
        return strokeCountVisible;
    }

    public int visibleStrokeCount() {
        return visibleStrokeCount;
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/HintVisibility.kt
package dev.bee.kanjianki.core.study

class HintVisibility(
    level: HintLevel?,
    private val tracePathsVisible: Boolean,
    private val outlineVisible: Boolean,
    private val strokeNumbersVisible: Boolean,
    private val startDotsVisible: Boolean,
    private val strokeCountVisible: Boolean,
    visibleStrokeCount: Int,
) {
    private val level: HintLevel = level ?: HintLevel.TRACE
    private val visibleStrokeCount: Int = maxOf(0, visibleStrokeCount)

    fun level(): HintLevel = level

    fun tracePathsVisible(): Boolean = tracePathsVisible

    fun outlineVisible(): Boolean = outlineVisible

    fun strokeNumbersVisible(): Boolean = strokeNumbersVisible

    fun startDotsVisible(): Boolean = startDotsVisible

    fun strokeCountVisible(): Boolean = strokeCountVisible

    fun visibleStrokeCount(): Int = visibleStrokeCount

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is HintVisibility) {
            return false
        }
        return level == other.level &&
            tracePathsVisible == other.tracePathsVisible &&
            outlineVisible == other.outlineVisible &&
            strokeNumbersVisible == other.strokeNumbersVisible &&
            startDotsVisible == other.startDotsVisible &&
            strokeCountVisible == other.strokeCountVisible &&
            visibleStrokeCount == other.visibleStrokeCount
    }

    override fun hashCode(): Int {
        var result = level.hashCode()
        result = 31 * result + tracePathsVisible.hashCode()
        result = 31 * result + outlineVisible.hashCode()
        result = 31 * result + strokeNumbersVisible.hashCode()
        result = 31 * result + startDotsVisible.hashCode()
        result = 31 * result + strokeCountVisible.hashCode()
        result = 31 * result + visibleStrokeCount
        return result
    }

    override fun toString(): String {
        return "HintVisibility[level=$level, tracePathsVisible=$tracePathsVisible, outlineVisible=$outlineVisible, " +
            "strokeNumbersVisible=$strokeNumbersVisible, startDotsVisible=$startDotsVisible, " +
            "strokeCountVisible=$strokeCountVisible, visibleStrokeCount=$visibleStrokeCount]"
    }
}
